README.rst
.. image:: https://img.shields.io/pypi/status/plone.app.fhirfield.svg
    :target: https://pypi.python.org/pypi/plone.app.fhirfield/
    :alt: Egg Status

.. image:: https://img.shields.io/travis/nazrulworld/plone.app.fhirfield/master.svg
    :target: http://travis-ci.org/nazrulworld/plone.app.fhirfield
    :alt: Travis Build Status

.. image:: https://coveralls.io/repos/github/nazrulworld/plone.app.fhirfield/badge.svg?branch=master
    :target: https://coveralls.io/github/nazrulworld/plone.app.fhirfield?branch=master
    :alt: Test Coverage

.. image:: https://img.shields.io/pypi/pyversions/plone.recipe.sublimetext.svg
    :target: https://pypi.python.org/pypi/plone.recipe.sublimetext/
    :alt: Python Versions

.. image:: https://img.shields.io/pypi/v/plone.app.fhirfield.svg
    :target: https://pypi.python.org/pypi/plone.app.fhirfield/
    :alt: Latest Version

.. image:: https://img.shields.io/pypi/l/plone.app.fhirfield.svg
    :target: https://pypi.python.org/pypi/plone.app.fhirfield/
    :alt: License

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/ambv/black


Background (plone.app.fhirfield)
================================

`FHIR`_ (Fast Healthcare Interoperability Resources) is the industry standard for Healthcare system. Our intend to implement `FHIR`_ based system using `Plone`_! `plone.app.fhirfield`_ will make life easier to create, manage content for `FHIR resources`_ as well search facilities for any FHIR Resources.

How It Works
------------

This field is working as like other `zope.schema <https://zopeschema.readthedocs.io/en/latest/>`_ field, just add and use it. You will feed the field value as either json string or python dict of `FHIR`_ resources through web form or any restapi client. This field has built-in `FHIR`_ resource validator and parser.

Example::

    from plone.app.fhirfield import FhirResource
    from plone.supermodel import model

    class IMyContent(model.Schema):

        <resource_type>_resource = FhirResource(
            title=u'your title',
            desciption=u'your desciption',
            fhir_release='any of FHIR release name'
            resource_type='any fhir resource type[optional]'
        )

The field's value is the instance of a specilized class `FhirResourceValue` inside the context, which is kind of proxy class of `fhir model <https://pypi.org/project/fhir.resources/>`_ with additional methods and attributes.


Features
--------

- Plone restapi support
- Widget: z3cform support
- plone.supermodel support
- plone.rfc822 marshaler field support

Available Field's Options
=========================

This field has got all standard options (i.e `title`, `required`, `desciption` and so on) with additionally options for the purpose of either validation or constraint those are related to `FHIR`_.


fhir_release
    Required: Yes

    Default: None

    Type: String

    The release version of `FHIR`_

    Example: ``R4``, ``STU3``


resource_type
    Required: No

    Default: None

    Type: String

    The name of `FHIR`_ resource can be used as constraint, meaning that we can restricted only accept certain resource. If FHIR json is provided that contains other resource type, would raise validation error.
    Example: `FhirResource(....,resource_type='Patient')`

model
    Required: No

    Default: None

    Type: String + full python path (dotted) of the model class.

    Like `resource_type` option, it can be used as constraint, however additionally this option enable us to use custom model class other than fhirclient's model.
    Example: `FhirResource(....,model='fhirclient.models.patient.Patient')`


index_mapping
    Required: No

    Default: None

    Type: JSON

    The custom index mapping, best case is elasticsearch mapping. Default mapping would be replaced by custom.

Field's Value API
=================

Field's value is a specilized class `plone.app.fhirfield.value.FhirResourceValue` which has reponsibilty to act as proxy of `fhirclient model's class <https://github.com/smart-on-fhir/client-py>`_. This class provides some powerful methods.

FhirResourceValue::json

    Originally this method is derived from fhir base model, you will always have to use this method during negotiation (although our serializer doing that for you automatically). This method not takes any argument, it returns FHIR json representation of resource.


FhirResourceValue::patch

    If you are familar with `FHIRPath Patch <https://www.hl7.org/fhir/fhirpatch.html>`_, this method one of the strongest weapon of this class. Patch applying on any `FHIR`_ resources is noting but so easy.
    This method takes one mandatory argument `patch_data` and that value should be list of patch items (`jsonpatch <http://jsonpatch.com/>`_).

    Example::

        from plone.app.fhirfield import FhirResource
        from plone.supermodel import model

        class ITask(model.Schema):

            resource = FhirResource(
                title=u'your title',
                desciption=u'your desciption',
                resource_type='Task'
            )

        patch_data = [
          {'op': 'replace', 'path': '/source/display', 'value': 'Patched display'},
          {'op': 'replace', 'path': '/status', 'value': 'Reopen'}
        ]
        task_content.resource.patch(patch_data)


FhirResourceValue::stringify

    This method returns string representation of fhir resource json value. Normally `as_json` returns python's dict type data. This method takes optional `prettify` argument, by setting this argument True, method will return human/print friendly representation.

FhirResourceValue::foreground_origin

    There may some situation come, where you will need just pure instance of fhir model, this method serves that purpose. This method returns current fhir resource model's instance.

    Example::

        from fhirclient.models.task import Task
        from plone.app.fhirfield import FhirResource
        from plone.supermodel import model

        class ITask(model.Schema):

            resource = FhirResource(
                title=u'your title',
                desciption=u'your desciption',
                resource_type='Task'
            )

        task = task_content.resource.foreground_origin()
        assert isinstance(task, Task)


Installation
============

Install plone.app.fhirfield by adding it to your buildout::

    [buildout]

    ...

    eggs =
        plone.app.fhirfield


and then running ``bin/buildout``. Go to plone control and install ``plone.app.fhirfield`` or If you are creating an addon that depends on this product, you may add ``<dependency>profile-plone.app.fhirfield:default</dependency>`` in ``metadata.xml`` at profiles.



Links
=====

Code repository:

    https://github.com/nazrulworld/plone.app.fhirfield

Continuous Integration:

    https://travis-ci.org/nazrulworld/plone.app.fhirfield

Issue Tracker:

    https://github.com/nazrulworld/plone.app.fhirfield/issues


License
=======

The project is licensed under the GPLv2.

.. _`FHIR`: https://www.hl7.org/fhir/overview.html
.. _`Plone`: https://www.plone.org/
.. _`FHIR Resources`: https://www.hl7.org/fhir/resourcelist.html
.. _`Plone restapi`: http://plonerestapi.readthedocs.io/en/latest/
.. _`plone.app.fhirfield`: https://pypi.org/project/plone.app.fhirfield/
.. _`jmespath`: https://github.com/jmespath/jmespath.py
.. _`jsonpath-rw`: http://jsonpath-rw.readthedocs.io/en/latest/
.. _`jsonpath-ng`: https://pypi.python.org/pypi/jsonpath-ng/1.4.3


CHANGES.rst
Changelog
=========

4.2.2 (unreleased)
------------------

- Nothing changed yet.


4.2.1 (2022-10-18)
------------------

Bugfixes

- `Issue-37: <https://github.com/nazrulworld/plone.app.fhirfield/issues/37>`_ Can not add a default value for fhirfield.


4.2.0 (2020-12-07)
------------------

Improvements

- Field deserializer is now accepting ``fhir.resources``'s model object, FhirResourceValue, alongside with string and dict type value.

- ``FhirResource::from_resource_model`` method has been added.

4.1.0 (2020-09-08)
------------------

Improvements

- Not any visibly improvement, but a lot of cleanup unused, unnecessary codes, in that impact made more slimmer version.

Breakings

- ``api`` module has been removed, use any import from ``plone.app.fhirfield`` instead.

- ``IFhirResourceModel`` interface is no longer available.

4.0.0 (2020-08-17)
------------------

Improvements

- supports `pydantic <https://pypi.org/project/pydantic/>`_ powered latest `fhir.resources <https://pypi.org/project/fhir.resources/>`_ via `fhirpath <https://pypi.org/project/fhirpath/>`_.
- suports Python 3.6.


Breakings

- Any FHIR search, FhirFieldIndex, EsFhirFieldIndex are removed. Please see `collective.fhirpath <https://pypi.org/project/collective.fhirpath/>`_ for those implementation.

- ``model_interface`` parameter has been removed from `FhirResource` field.

- Removed three ``fhirfield.es.index.mapping`` registry.

- extra option ``elasticsearch`` has been removed from setup.py


3.1.1 (2020-05-15)
------------------

- Nothing changed but some documents updated.


3.1.0 (2020-05-14)
------------------

Breakings

- Drop support python version until 3.6.x

- ``FhirResource.get_fhir_version`` been changed to ``FhirResource.get_fhir_release`` and no longer returns Enum member instead just string value.

- One of required ``FhirResource`` init parameter ``fhir_version`` has been changed to ``fhir_release``.

Removed Helper Functions

- ``plone.app.fhirfield.helpers.resource_type_str_to_fhir_model`` use ``fhirpath.utils.lookup_fhir_class``.

- ``plone.app.fhirfield.helpers.search_fhir_model`` use ``fhirpath.utils.lookup_fhir_class_path``

- ``plone.app.fhirfield.helpers.import_string`` use ``fhirpath.utils.import_string``


3.0.0 (2020-03-18)
------------------

Improvements

- Elasticsearch mappings are updated with correct value.


3.0.0b2 (2019-11-12)
--------------------

Newfeatures

- Fully compatiable with `collective.fhirpath`_.

- It is possible to provide custom index mapping through ``FhirResource`` field.

Breakings

- ``fhir_version`` value is manadatory, so you have provide fhir version number.

Deprecations

- Using PluginIndexes (EsFhirFieldIndex) has been deprecated.

- FHIR Search through ``portal_catalog`` has been deprecated.

- Using elasticsearch mapping from this project has been deprecated.


3.0.0b1 (2019-10-01)
--------------------

Newfeatures

- `FHIRPath`_ support added through `collective.fhirpath`_.

- Now supports Elasticsearch server version ``6.8.3``.


Breakings

- Drop support python 2.x.x.

2.0.0 (2019-05-18)
------------------

Newfeatures

- `Issue#14 <https://github.com/nazrulworld/plone.app.fhirfield/issues/14>`_ Now reference query is powerful yet!
  It is possible now search by resourceType only or mixing up.

- `Issue#21 <https://github.com/nazrulworld/plone.app.fhirfield/issues/21>`_ One of the powerful feature added, IN/OR support in search query.
  Now possible to provide multiple values separated by comma.

Breakings

- Drop support Elastic server version 2.3.x.


Bugfixes

- Important fix for ``Quantity`` search type, now value prefix not impact on other (unit, system). Additionally
  also now possible to search by unit or system and code (without value)


1.0.0 (2019-01-18)
------------------

Breaking

- Drop ``fhirclient`` dependency, instead make ``fhir.resources`` dependency.
- ``collective.elasticsearch`` version minimum version ``2.0.5`` has been pinned.


1.0.0rc4 (2018-10-25)
---------------------

Breaking

- Drop support for Plone 4.xx (sorry).

Improvements

- Issue#10 JSON Viewer added in display mode.

- Issue#18 `api` module to make available for all publicly usable methods, functions, classes.

- Issue#17 Add suports for duplicate param names into query string. It is now possible to provide multiple condition for same param type. For example `?_lastUpdated=gt2015-10-15T06:31:18+00:00&_lastUpdated=lt2018-01-15T06:31:18+00:00`

- Issue#10 Add support for `Composite` type FHIR search param.

- Issue#13 Add support for `Address` and `ContactPoint` mapping. It opens up many search params.

- Mappings are more enriched, so many missing mappings are added, like `valueString`, `valueQuantity` and so on.[nazrulworld]

- Issue#12 Full support for `code` search param type has been added, it also opens up for other search parameters (y).

- Issue#15 support for `Humanane` mapping in search.


1.0.0rc3 (2018-09-22)
---------------------

Improvements

- `Issue: 6 <https://github.com/nazrulworld/plone.app.fhirfield/issues/6>`_ A major improvement has been done, now very slim version (`id`, `resourceType`, `meta`) of FHIR json resource has been indexed in ZCatalog (zope index) version, however previously whole fhir resource was stored as a result now huge storage savings, perhaps improves indexing performance. [nazrulworld]


1.0.0rc2 (2018-08-29)
---------------------

Bugfixes

- Issue 5: `FHIR search's modifier 'missing' is not working for nested mapping <https://github.com/nazrulworld/plone.app.fhirfield/issues/5>`_

1.0.0rc1 (2018-08-27)
---------------------

Newfeatures

- `Identifier search parameter <http://www.hl7.org/fhir/search.html#token>`_ is active now (both array of identifier and single object identifier).

- Array of Reference query support (for example `basedOn` (list of reference) ) is active now. Although normal object reference has already been supported.

- All available mappings for searchable resources are generated.


scripts/mappings.py
# -*- coding: utf-8 -*-
# @Date    : 2018-08-21 17:48:31
# @Author  : Md Nazrul Islam <email2nazrul@gmail.com>
# @Link    : http://nazrul.me/
# @Version : $Id$
# All imports here
"""Compound mapping types """

__author__ = 'Md Nazrul Islam <email2nazrul@gmail.com>'

date_pattern = '-?[0-9]{4}(-(0[1-9]|1[0-2])(-(0[0-9]|[1-2][0-9]|3[0-1]))?)?'
datetime_pattern = '-?[0-9]{4}(-(0[1-9]|1[0-2])(-(0[0-9]|[1-2][0-9]|3[0-1])(T([01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\\.[0-9]+)?(Z|(\\+|-)((0[0-9]|1[0-3]):[0-5][0-9]|14:00)))?)?)?'
Bool = {
    "type": "boolean",
    "store": False
}
Float = {
    "type": "float",
    "store": False
}

Integer = {
    "type": "integer",
    "store": False
}
Long = {
    "type": "long",
    "store": False
}

Token = {
    'type': 'keyword',
    'index': True,
    'store': False
}

KeywordToken = {
    'type': 'keyword',
    'index': True,
    'store': False
}

PathToken = {"type": "text", "index": True, "store": False, "analyzer": "path_analyzer"}

Text = {
    'type': 'text',
    'index': True,
    'store': False,
    'analyzer': 'standard'
}

SearchableText = {
    'type': 'text',
    'index': True,
    'analyzer': 'standard',
    'store': False
}

Date = {
    'type': 'date',
    'format': 'date_time_no_millis||date_optional_time',
    'store': False
}

Timing = {
    'properties': {
        'event': Date,
        'code': Token
    }
}

Reference = {
    "properties": {
        "reference": {
            "type": "text",
            "index": True,
            "store": False,
            "analyzer": "fhir_reference_analyzer",
        }
    }
}

Attachment = {
    'properties': {
        'url': Token,
        'language': Token,
        'title': Text,
        'creation': Date
    }
}

Coding = {
    'properties': {
        'system': Token,
        'code': Token,
        'display': Token
    }
}

CodeableConcept = {
    'properties': {
        'text': Text,
        'coding': {
            'type': 'nested',
            'properties': Coding['properties']
        }
    }
}

Period = {
    'properties': {
        'start': Date,
        'end': Date
    }
}
Identifier = {
    'properties': {
        'use': Token,
        'system': Token,
        'value': Token,
        'type': {
            'properties': {
                'text': Text
            }
        }
    }
}
Quantity = {
    'properties': {
        'value': Float,
        'code': Token,
        'system': Token,
        'unit': Token
    }
}

Money = Quantity
Range = {
    'properties': {
        'high': Quantity,
        'low': Quantity
    }
}

Age = Quantity
Address = {
    'properties': {
        'city': Token,
        'country': Token,
        'postalCode': Token,
        'state': Token,
        'use': Token
    }
}

HumanName = {
    'properties': {
        'family': Token,
        'text': Text,
        'prefix': Token,
        'given': Token,
        'use': Token
    }
}
Duration = Quantity

ContactPoint = {
    'properties': {
        'period': Period,
        'rank': Integer,
        'system': Token,
        'use': Token,
        'value': Text
    }
}

ContactDetail = {
    'properties': {
        'name': Token,
        'telecom': ContactPoint
    }
}

ContactDetail['properties']['telecom'].update({
    'type': 'nested'
})

Annotation = {
    'properties': {
        'authorReference': Reference,
        'authorString': Text,
        'text': Text,
        'time': Date
    }
}

Dosage = {
    'properties': {
        'asNeededBoolean': Bool,
        'asNeededCodeableConcept': CodeableConcept,
        'doseQuantity': Quantity,
        'doseRange': Range,
        'site': CodeableConcept,
        'text': Text,
        'timing': Timing
    }
}

RelatedArtifact = {
    'properties': {
        'type': Token,
        'url': Token,
        'resource': Reference
    }
}

Signature = {
    'properties': {
        'contentType': Token,
        'when': Date,
        'whoReference': Reference,
        'whoUri': Token
    }
}
# Common
Id = Token
Meta = {
    'properties': {
        'versionId': Token,
        'lastUpdated': Date,
        'profile': Token
    }
}
